### ISSUE TITLE
Controlling user right 'smw-admin' does not work as expected

### ISSUE THREAD

**Issue by kghbln** [SemanticMediaWiki/SemanticMediaWiki#1137]
**Default:**
```text
$wgGroupPermissions['sysop']['smw-admin'] = true;
$wgGroupPermissions['smwadministrator']['smw-admin'] = true;
```
*Result:*
Users in the 'sysop' and 'smwadministrator' group may access "Special:SWM-Admin" --> ok

**Deviation 1:**
```text
$wgGroupPermissions['sysop']['smw-admin'] = false;
$wgGroupPermissions['smwadministrator']['smw-admin'] = true;
```
*Result:*
Users in the 'sysop' and 'smwadministrator' group may access "Special:SWM-Admin". -->  not ok, since a person only in the 'sysop' group may still access "Special:SMW-Admin".

**Deviation 2:**
```text
$wgGroupPermissions['sysop']['smw-admin'] = false;
$wgGroupPermissions['smwadministrator']['smw-admin'] = false;
```
*Result:*
Users in the 'sysop' and 'smwadministrator' group may access "Special:SWM-Admin". -->  not ok, since a person either in the 'sysop' or 'smwadministrator' group may still access "Special:SMW-Admin".

The only solution to prevent access to "Special:SMW-Admin" currently is to set
```text
$wgRevokePermissions['sysop']['smw-admin'] = true;
$wgGroupPermissions['smwadministrator']['smw-admin'] = true;
```
The problem is that a user cannot be in the 'sysop' group and the 'smwadministrator' at the same time since `$wgRevokePermissions` takes precedence.

Tested with a1e7e43 on MW 1.24.3 using PHP 5.41. See also the [thread on smw-user](http://wikimedia.7.x6.nabble.com/SMW-Question-smw-admin-rights-td5052285.html).

**Comment by mwjames** [comment]
The lines that define the permission can be found at [0] with check being executed in [1]. Feel free to sent a PR (I'm not going to look into this issue).

[0] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/includes/Setup.php#L160-L164

[1] https://github.com/SemanticMediaWiki/SemanticMediaWiki/blob/master/includes/specials/SMW_SpecialSMWAdmin.php#L41

**Comment by mwjames** [comment]
Try something like this to prevent `LocalSettings` being reset to a default setting.

```
-		$this->globalVars['wgGroupPermissions']['sysop']['smw-admin'] = true;
-		$this->globalVars['wgGroupPermissions']['smwadministrator']['smw-admin'] = true;
+		if ( !isset( $this->globalVars['wgGroupPermissions']['sysop']['smw-admin'] ) ) {
+			$this->globalVars['wgGroupPermissions']['sysop']['smw-admin'] = true;
+		}
+
+		if ( !isset( $this->globalVars['wgGroupPermissions']['smwadministrator']['smw-admin'] ) ) {
+			$this->globalVars['wgGroupPermissions']['smwadministrator']['smw-admin'] = true;
+		}
```

**Comment by JeroenDeDauw** [comment]
Yeah, presumably something is going wrong that causes us to override the user specified config with the default. If that is the case, it might be going wrong for other settings as well.

**Comment by mwjames** [comment]
> Yeah, presumably something is going wrong that causes us to override the user specified config with the default. If that is the case, it might be going wrong for other settings as well.

It is just that since we registered late in `$GLOBALS['wgExtensionFunctions']` (to ensure that LocalSettings are loaded before) you have to check whethersettings have been already set (as shown above) or not.

**Comment by mwjames** [comment]
#233 being of the same cause with a similar solution.

**Comment by JeroenDeDauw** [comment]
This problem was created by one of us, or perhaps both of us.

**Comment by mwjames** [comment]
> This problem was created by one of us, or perhaps both of us.

Not sure what you referring to but as we started to use Composer autoloading, `$GLOBALS['wgExtensionFunctions']` was the only way to make sure that settings are done before starting to inject SMW related code. Furthermore, we isolated the usage of `$GLOBALS` to avoid that user can come by and change settings freely.

I already gave an answer has to how to fix this issue.
